mpa/cls/explainer.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (C) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+import os.path as osp
+import torch
+
+import mmcv
+from mmcv.parallel import MMDataParallel
+from mmcv.runner import load_checkpoint, wrap_fp16_model
+
+from mmcls.datasets import build_dataloader, build_dataset
+from mmcls.models import build_classifier
+
+from mpa.registry import STAGES
+from mpa.cls.stage import ClsStage
+from mpa.modules.hooks.recording_forward_hooks import ActivationMapHook, EigenCamHook
+from mpa.utils.logger import get_logger
+logger = get_logger()
+EXPLAINER_HOOK_SELECTOR = {
+    'eigencam': EigenCamHook,
+    'activationmap': ActivationMapHook}
+
+
+@STAGES.register_module()
+class ClsExplainer(ClsStage):
+    def run(self, model_cfg, model_ckpt, data_cfg, **kwargs):
+        """Run explain stage
+        - Configuration
+        - Environment setup
+        - Run explain via recording_forward_hooks
+        """
+        self._init_logger()
+        mode = kwargs.get('mode', 'train')
+        if mode not in self.mode:
+            return {}
+        explainer = kwargs.get('explainer')
+        self.explainer_hook = EXPLAINER_HOOK_SELECTOR.get(explainer.lower(), None)
+        if self.explainer_hook is None:
+            raise NotImplementedError(f"explainer algorithm {explainer} not supported")
+        logger.info(f"explainer algorithm: {explainer}")
+        cfg = self.configure(model_cfg, model_ckpt, data_cfg, training=False, **kwargs)
+
+        mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
+        outputs = self._explain(cfg)
+        return dict(outputs=outputs)
+
+    def _explain(self, cfg):
+        if cfg.get('task_adapt', False) and not hasattr(self, 'eval'):
+            dataset_cfg = cfg.data.train.copy()
+            dataset_cfg.pipeline = cfg.data.test.pipeline
+            self.dataset = build_dataset(dataset_cfg)
+        else:
+            self.dataset = build_dataset(cfg.data.test)
```

Comment:
Could you refactor this block to a method with proper name?

Reply:
Fixed to `explain_dataset` and removed if - else part